### PR TITLE
[Doc] Fix broken references in Ray Core documentation

### PR DESCRIPTION
**Pull request by peytondmurray**
## Why are these changes needed?

This PR fixes broken links in Ray Core documentation in preparation for turning on Sphinx's nitpicky mode.

## Related issue number

Partially addresses #39658. Blocked until #45160 is merged.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/ray-core/api/utility.rst [hunk 1]
```diff
@@ -10,6 +10,8 @@ Utility
    ray.nodes
    ray.cluster_resources
    ray.available_resources
+   ray.util.queue.Empty
```
Comment (can-anyscale) [inline doc/source/ray-core/api/utility.rst:13]: can you comment inline why we're adding these to the doc, thankks

File: doc/source/ray-more-libs/dask-on-ray.rst [hunk 2]
```diff
@@ -224,8 +224,7 @@ The following Ray-specific callbacks are provided:
    6. :code:`ray_finish(result)`: Run after all Ray tasks have finished
       executing and the final result has been returned.
 
-See the docstring for
-:meth:`RayDaskCallback.__init__() <ray.util.dask.callbacks.RayDaskCallback>.__init__`
+See the docstring for :meth:`~ray.util.dask.RayDaskCallback.__init__`
```
Comment (can-anyscale) [inline doc/source/ray-more-libs/dask-on-ray.rst:227]: qq, what the `~` means?

File: doc/source/ray-more-libs/dask-on-ray.rst [hunk 3]
```diff
@@ -272,14 +271,18 @@ execution time exceeds some user-defined threshold:
...
    :toctree: doc/
 
-   ray.util.dask.callbacks.RayDaskCallback
-   ray.util.dask.callbacks.RayDaskCallback._ray_presubmit
-   ray.util.dask.callbacks.RayDaskCallback._ray_postsubmit
-   ray.util.dask.callbacks.RayDaskCallback._ray_pretask
-   ray.util.dask.callbacks.RayDaskCallback._ray_posttask
-   ray.util.dask.callbacks.RayDaskCallback._ray_postsubmit_all
-   ray.util.dask.callbacks.RayDaskCallback._ray_finish
+   ray.util.dask.RayDaskCallback
```
Comment (can-anyscale) [inline doc/source/ray-more-libs/dask-on-ray.rst:281]: this class is inside the `callbacks.py` module; why do we remove the `callbacks` part?

File: python/ray/_private/worker.py [hunk 4]
```diff
@@ -2708,14 +2708,14 @@ def put(
 @PublicAPI
 @client_mode_hook
 def wait(
-    ray_waitables: List[Union["ObjectRef[R]", "ObjectRefGenerator[R]"]],
+    ray_waitables: List[Union[ObjectRef, ObjectRefGenerator]],
```
Comment (can-anyscale) [inline python/ray/_private/worker.py:2711]: what is `R` and why we remove it?

**Review by can-anyscale — COMMENTED** [review]
thanks for the clean up, a bunch of dump questions mostly for me to learn

File: doc/source/ray-more-libs/dask-on-ray.rst (hunk 2 above)
Comment (peytondmurray) [reply doc/source/ray-more-libs/dask-on-ray.rst:227]: It tells Sphinx to not include the `ray.util.dask.RayDaskCallback` part when the link is rendered. So the link will appear as just `__init__()` in the docs.

File: doc/source/ray-more-libs/dask-on-ray.rst (hunk 3 above)
Comment (peytondmurray) [reply doc/source/ray-more-libs/dask-on-ray.rst:281]: This has to do with the way that `sphinx.ext.autosummary` resolves imports to the most base module where it is exported. Here, `RayDaskCallback` is explicitly exported as part of `ray.util.dask.__init__.py`: https://github.com/ray-project/ray/blob/18acf026d23c6520523e12a31d251773fee00c54/python/ray/util/dask/__init__.py#L57. The link won't resolve otherwise.

File: python/ray/_private/worker.py (hunk 4 above)
Comment (peytondmurray) [reply python/ray/_private/worker.py]: This is [type parameter syntax](https://docs.python.org/3/reference/compound_stmts.html#type-params). There's information [here](https://docs.python.org/3/library/typing.html#generics) on how it works, but this is a way of parameterizing a type using a generic. These forward references won't resolve unless the types are defined or imported here, so that's why they're changed to concrete references.

File: doc/source/ray-core/api/utility.rst (hunk 1 above)
Comment (peytondmurray) [reply doc/source/ray-core/api/utility.rst:13]: Sure, but it's because they're referred to elsewhere in the documentation.

**Review by can-anyscale — APPROVED** [review]

File: python/ray/_private/worker.py [hunk 5]
```diff
@@ -2708,14 +2708,14 @@ def put(
 @PublicAPI
 @client_mode_hook
 def wait(
-    ray_waitables: List[Union["ObjectRef[R]", "ObjectRefGenerator[R]"]],
+    ray_waitables: Union[ObjectRef, ObjectRefGenerator],
```
Comment (jjyao) [inline python/ray/_private/worker.py:2711]: This should be a list.

Why changing this? Seems unrelated to broken referneces

File: python/ray/util/state/api.py [hunk 6]
```diff
@@ -1468,8 +1449,7 @@ def summarize_objects(
         Dictionarified :class:`~ray.util.state.common.ObjectSummaries`
 
     Raises:
-        Exceptions: :class:`RayStateApiException <ray.util.state.exception.RayStateApiException>` if the CLI
-            failed to query the data.
+        RayStateApiException: if the CLI failed to query the data.
```
Comment (jjyao) [inline python/ray/util/state/api.py:1452]: Does this generate the link?

**Review by jjyao — CHANGES_REQUESTED** [review]

File: python/ray/_private/worker.py (hunk 5 above)
Comment (peytondmurray) [reply python/ray/_private/worker.py:2711]: Sorry, my mistake - I'll make this a list again.

> Why changing this? Seems unrelated to broken referneces

`sphinx.ext.autodoc` uses type annotations to document parameters and return types for functions.

File: python/ray/util/state/api.py (hunk 6 above)
Comment (peytondmurray) [reply python/ray/util/state/api.py:1452]: Yes.

File: python/ray/_private/worker.py [hunk 7]
```diff
@@ -2708,14 +2708,14 @@ def put(
...
-    ray_waitables: List[Union["ObjectRef[R]", "ObjectRefGenerator[R]"]],
+    ray_waitables: List[Union[ObjectRef, ObjectRefGenerator]],
     *,
     num_returns: int = 1,
     timeout: Optional[float] = None,
     fetch_local: bool = True,
 ) -> Tuple[
-    List[Union["ObjectRef[R]", "ObjectRefGenerator[R]"]],
-    List[Union["ObjectRef[R]", "ObjectRefGenerator[R]"]],
+    List[Union[ObjectRef, ObjectRefGenerator]],
```
Comment (jjyao) [inline python/ray/_private/worker.py:2717]: Is `R` removed so that we can generate link to ObjectRef?

Comment (peytondmurray) [reply python/ray/_private/worker.py:2717]: Yes, although if you'd prefer to keep `R`, we can import it at the top and instead have these be `ObjectRef[R]`. So depending on how you'd like to handle this, we can either keep it as is or add the `TypeVar`.

Comment (jjyao) [reply python/ray/_private/worker.py:2717]: @rkooo567 what do you want here since you added `R` I think

**Review by jjyao — APPROVED** [review]
Thanks for the fix.
